### PR TITLE
Statistics plugin v2 (SQLite)

### PR DESCRIPTION
**Pull request by robert00s**
Discussion: #2378 
For now I used events (UpdateStats) to send average time per page to ReaderFooter. I know that It's polluting the core code path with plugin specific logic but I don't time and knowledge to create footer module fully pluggable.
The conversion is done at the first KOreader startup after the update
We have info:
zrzut01

and after conversion:
zrzut02

There are difference between time in new and old statistics up to 5% after conversion because algorithm was changed and this #2836.

### CODE REVIEW

File: plugins/statistics.koplugin/main.lua [hunk 1]
```diff
@@ -73,58 +83,350 @@ function ReaderStatistics:init()
...
+    return m:sum()
+end
+
+function ReaderStatistics:createDB()
+    local conn = SQ3.open(db_location)
+    conn:exec("CREATE TABLE IF NOT EXISTS book(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT, authors TEXT," ..
+        "notes INTEGER, last_open INTEGER, highlights INTEGER, pages INTEGER, series TEXT, language TEXT," ..
+        "md5 TEXT, total_read_time INTEGER, total_read_pages INTEGER);" ..
+        "CREATE TABLE IF NOT EXISTS page_stat(id_book INTEGER, page INTEGER NOT NULL, start_time INTEGER NOT NULL, " ..
+        "period INTEGER NOT NULL, UNIQUE (page, start_time), FOREIGN KEY(id_book) REFERENCES book(id));" ..
```
Comment (apletnev) [inline plugins/statistics.koplugin/main.lua]: why do you use a "period" field if you have a "start_time" field? Before was start_time(n) - start_time(n-1) is period? Or am I wrong?

File: plugins/statistics.koplugin/main.lua [hunk 2]
```diff
@@ -284,321 +600,264 @@ function ReaderStatistics:addToMainMenu(menu_items)
...
-        self.pages_current_period = self.pages_current_period + 1
-        self.data.total_time_in_sec = self.data.total_time_in_sec + diff_time
-        self.data.performance_in_pages[curr_time.sec] = pageno
-        -- we cannot save stats each time this is a page update event,
-        -- because the self.data may not even be initialized when such a event
-        -- comes, which will render a blank stats written into doc settings
-        -- and all previous stats are totally wiped out.
-        self.ui.doc_settings:saveSetting("stats", self.data)
+    -- every 50 pages we write stats to database
+    if util.tableSize(self.pages_stats) % 50 == 0 then
```
Comment (apletnev) [inline plugins/statistics.koplugin/main.lua]: Make a number (50) as constant.

File: plugins/statistics.koplugin/main.lua [hunk 3]
```diff
@@ -73,58 +83,350 @@ function ReaderStatistics:init()
...
+    end
+    table.sort(sorted_performance)
+    conn:exec('BEGIN')
+    local stmt = conn:prepare("INSERT OR IGNORE INTO page_stat VALUES(?, ?, ?, ?)")
+    for i=1, #sorted_performance - 1 do
+        diff_time = sorted_performance[i+1] - sorted_performance[i]
+        if diff_time <= self.page_max_read_sec and diff_time >= self.page_min_read_sec  then
+            stmt:reset():bind(id_book, self.pages_stats[sorted_performance[i]], sorted_performance[i], diff_time):step()
+        elseif diff_time > self.page_max_read_sec then
+            stmt:reset():bind(id_book, self.pages_stats[sorted_performance[i]], sorted_performance[i], self.page_max_read_sec):step()
```
Comment (apletnev) [inline plugins/statistics.koplugin/main.lua]: Why do you save a value (page_max_read_sec) in case when a user did not read a page for page_max_read_sec minutes? If a user did not read a page, then, he will reread the same page or skip it. In this case, when user reread the same page you will add incorrect value page_max_read_sec and an actual correct read value.

File: plugins/statistics.koplugin/main.lua [hunk 4]
```diff
@@ -73,58 +83,350 @@ function ReaderStatistics:init()
...
+
+function ReaderStatistics:insertDB(id_book)
+    self.pages_stats[TimeVal:now().sec] = self.curr_page
+    if id_book == nil or util.tableSize(self.pages_stats) < 2 then
+        return
+    end
+    local diff_time
+    local conn = SQ3.open(db_location)
+    local sorted_performance = {}
+    for k, v in pairs(self.pages_stats) do
```
Comment (apletnev) [inline plugins/statistics.koplugin/main.lua]: You can avoid sorting the table, in case when you calculate difference between pages if a diff will be < 0  you can just multiply it to -1

File: plugins/statistics.koplugin/main.lua [hunk 5]
```diff
@@ -1,32 +1,41 @@
...
+local db_location = DataStorage:getSettingsDir() .. "/book.sqlite3"
 
 local ReaderStatistics = Widget:extend{
-    last_time = nil,
-    page_min_read_sec = 5,
-    page_max_read_sec = 90,
-    current_period = 0,
-    pages_current_period = 0,
+    page_min_read_sec = 1,
+    page_max_read_sec = 180,
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: I timed it. On this page with difficult French 60s is only enough to finish the top two paragraphs. Now I'm more interested in when there was reading activity than accurate reading time stats but clearly depending on what and how you're reading 3 minutes isn't some outrageous value.

p_20170514_211611

Comment (AlanSP1) [inline plugins/statistics.koplugin/main.lua]: Well, I have two things on this:

First, shouldn't this new statistics use old user settings for statistics? We can define (or we could before) our timings.

Second, time to read page depends very much on many things. I personally want smallest font I can see (with time, it surely increase in size) and if you have bigger reader (i.e. bigger screen), that can take some time. For me, at the moment, my average page takes about 3 minutes.

In this case, both values are "wrong" for me. My minimum is 15, maximum is 900. I think it would be best if timings are user customizable.

Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Quite true. However, my point is about sensible defaults. 1s obviously isn't, but I don't think 60s is either.

Comment (AlanSP1) [inline plugins/statistics.koplugin/main.lua]: Defaults for old statistics module were 5 and 90 if I'm not mistaken.

Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Better, but if that was the default it's still too short because it caused bug reports about stats not working. The default should then be something more like two or three minutes. Even if you did put down your reader for a minute and "forgot", is it really so bad if the device counts an extra minute occasionally?

I can totally see where you're coming from with 15 minutes when you're writing down your thoughts on what you're reading, looking things up in reference works, double-checking calculations, etc., but that wouldn't do as a default. ;-)

Comment (AlanSP1) [inline plugins/statistics.koplugin/main.lua]: Of course, these are my personal settings. For defaults, I maybe would go with 10 and 300.

File: plugins/statistics.koplugin/main.lua [hunk 6]
```diff
@@ -1,32 +1,41 @@
...
+local statistics_dir = DataStorage:getDataDir() .. "/statistics/"
+local db_location = DataStorage:getSettingsDir() .. "/book.sqlite3"
 
 local ReaderStatistics = Widget:extend{
-    last_time = nil,
-    page_min_read_sec = 5,
-    page_max_read_sec = 90,
-    current_period = 0,
-    pages_current_period = 0,
+    page_min_read_sec = 1,
```
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: It's only default value. Statistics v2 don't change users settings:
https://github.com/robert00s/koreader/blob/07e8b2dc6f79443b75d2e820b14733e0d82d5e63/plugins/statistics.koplugin/main.lua#L88

File: plugins/statistics.koplugin/main.lua (hunk 5 above)
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: It's only default value. Every user can change this value in settings.
Statistics v2 don't change this value.
It's only suggested (default) value for new users.

File: plugins/statistics.koplugin/main.lua [hunk 7]
```diff
@@ -73,58 +83,350 @@ function ReaderStatistics:init()
...
-end
+    self.convert_to_db = settings.convert_to_db
+    self.ui.menu:registerToMainMenu(self)
 
-function ReaderStatistics:getBookProperties()
-    local props = self.view.document:getProps()
-    if props.title == "No document" or props.title == "" then
-        -- FIXME: sometimes crengine returns "No document", try one more time
-        props = self.view.document:getProps()
+    if self.convert_to_db then      -- if conversion to sqlite was doing earlier
```
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: I will change it soon.

File: plugins/statistics.koplugin/main.lua (hunk 1 above)
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: I create period because it's much easier and faster to make SQL statements for statistic like "Today Period", "Current Period", Time in Months, Weeks etc.
For example - sum(period):

```
SELECT sum(period), count(DISTINCT page) 
FROM page_stat 
WHERE id_book = 'id'
```

File: plugins/statistics.koplugin/main.lua (hunk 2 above)
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: I will change it soon.

File: plugins/statistics.koplugin/main.lua (hunk 3 above)
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: It is difficult to say whether a user has not read this page or read more than the maximum time.
I think we can not ignore it and skip this page.

File: plugins/statistics.koplugin/main.lua (hunk 4 above)
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: Thanks, I'll check it.

File: plugins/statistics.koplugin/main.lua (hunk 3 above)
Comment (apletnev) [inline plugins/statistics.koplugin/main.lua]: To avoid this misunderstanding, I think we need to add something like "book opened" feature. And provide info, for example, the book was open for an 1 hour, approximately read for 50 minutes. What do you think? In this case, you will count reading as the process (what do you want), and for guys who want improve their reading or get details (like me) they can receive detailed statistics by pages without their sleeping process ;).

File: plugins/statistics.koplugin/main.lua (hunk 5 above)
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: I think 180 is an acceptable default, but the idea that it's "only" a default is a very dangerous one that I vehemently oppose.

File: plugins/statistics.koplugin/main.lua (hunk 6 above)
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: But why did you change it from the current default value of `5`?

Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: I changed it for tests.
I restore it to previous values.

File: plugins/statistics.koplugin/main.lua (hunk 4 above)
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: Unfortunately the array must be sorted.
We need to save the pages to the database in order of reading and calculate the time for each page.

File: plugins/statistics.koplugin/main.lua (hunk 3 above)
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: I can think about it in statistics v2.1 version :)

**Comment by robert00s** [comment]
Maybe `150`? :)

**Comment by KenMaltby** [comment]
There needs to be an option to totally disable any interaction with the reader loop.  The "plugin" should not be allowed to impact the reader performance or expose it to plugin induced error or crash.  If necessary, most of us can find and disable the appropriate code, but the average user, who just wants a reader, should have an effective menu option.  An option to opt out of any of this statistics gathering processing, would be greatly appreciated.

**Comment by robert00s** [comment]
@KenMaltby 
You can disable statistics by uncheck `Reading statistics -> Enable` option.

**Comment by houqp** [comment]
Nice work @robert00s :+1: Thanks @apletnev for the reviews! I also will help review this later this week.

File: plugins/statistics.koplugin/main.lua [hunk 8]
```diff
@@ -284,321 +604,264 @@ function ReaderStatistics:addToMainMenu(menu_items)
...
+    if ptype == "daily" or ptype == "daily_weekday" then
+        modifier = '%Y-%m-%d'
+    elseif ptype == "weekly" then
+        modifier = '%Y-%W'
+    elseif ptype == "monthly" then
+        modifier = '%Y %m'
+    end
+    self:insertDB(self.id_curr_book)
+    local conn = SQ3.open(db_location)
+    local result_book = conn:exec(string.format("SELECT strftime('%s', start_time, 'unixepoch', 'localtime') AS dates, " ..
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Again, anything more complex I think would be really improved by making it a multiline SQL query string. Note that my suggestions are just about the general principle of the thing and offer no concrete statement about the specifics of formatting other than that having some really improves clarity.

```lua
[[SELECT Strftime('%s', start_time, 'unixepoch', 'localtime') AS dates,
    Count(DISTINCT page)                                      AS pages,
    Sum(period)                                               AS periods,
    start_time
FROM   page_stat
WHERE  start_time >= '%s'
GROUP  BY dates
ORDER  BY dates DESC;]]
```

File: plugins/statistics.koplugin/main.lua [hunk 9]
```diff
@@ -284,321 +604,264 @@ function ReaderStatistics:addToMainMenu(menu_items)
...
+        "count(distinct page) AS pages, sum(period) AS periods, start_time FROM page_stat WHERE start_time >= '%s' " ..
+        "GROUP BY dates ORDER BY dates desc;", modifier, period_begin))
+    local result_pages = conn:exec(string.format("SELECT strftime('%s', start_time, 'unixepoch', 'localtime') AS dates, " ..
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: This one too. I've used http://www.dpriver.com/pp/sqlformat.htm as a quick formatting aid, hence initial caps on the function names which is the default there.

File: plugins/statistics.koplugin/main.lua [hunk 10]
```diff
@@ -73,58 +84,353 @@ function ReaderStatistics:init()
...
+        else
+            break
+        end
+    end
+    return m:sum()
+end
+
+function ReaderStatistics:createDB()
+    local conn = SQ3.open(db_location)
+    conn:exec("CREATE TABLE IF NOT EXISTS book(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT, authors TEXT," ..
```
Comment (robert00s) [inline plugins/statistics.koplugin/main.lua]: What do you think about something like that:

```
function ReaderStatistics:createDB()
    local conn = SQ3.open(db_location)
    local sql_stmt = [[
        CREATE TABLE IF NOT EXISTS book
            (
                id integer PRIMARY KEY autoincrement,
                title text,
                authors text,
                notes      integer,
                last_open  integer,
                highlights integer,
                pages      integer,
                series text,
                language text,
                md5 text,
                total_read_time  integer,
                total_read_pages integer
            );
        CREATE TABLE IF NOT EXISTS page_stat
            (
                id_book    integer,
                page       integer NOT NULL,
                start_time integer NOT NULL,
                period     integer NOT NULL,
                UNIQUE (page, start_time),
                FOREIGN KEY(id_book) REFERENCES book(id)
             );
        CREATE TABLE IF NOT EXISTS info
             (
                 version integer
             ); 
    ]]
    conn:exec(sql_stmt)
    --DB structure version - now is version 1
    local stmt = conn:prepare("INSERT INTO info values (?)")
    stmt:reset():bind("1"):step()
    stmt:close()
    conn:close()
end
```

Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Yeah, that's exactly what I had in mind. 👍

File: plugins/statistics.koplugin/main.lua [hunk 11]
```diff
@@ -369,236 +789,282 @@ end
...
-                --add average time if time > page_max_time e.g longer break while reading
-                entry.read = book.total_time_in_sec / read_pages + entry.read
-            end
-            entry.date = n
-            entry.count = entry.count + 1
+            date_text = result_book[1][i]
         end
+        table.insert(results, {
+            date_text,
+            T(_("Pages (%1) Time: %2"), tonumber(result_pages[2][i]), util.secondsToClock(tonumber(result_book[3][i]), false))
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Is that supposed to have a colon after pages like this?
```lua
"Pages: (%1) Time: %2"
```

**Comment by Frenzie** [comment]
It's lovely to read now. Previously those SQL statements were basically a wall of text and now I can just glance at them to see what they're doing. A big boon to future maintainability. :-)

File: plugins/statistics.koplugin/main.lua [hunk 12]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
+        local book_title = book_stats.title:gsub("'", "''")
+        local book_authors = book_stats.authors:gsub("'", "''")
+        sql_stmt = [[
+            SELECT count(id)
+            FROM   book
+            WHERE  title = '%s'
+                AND    authors = '%s'
+                AND    md5 = '%s';
+        ]]
+        res = conn:rowexec(string.format(sql_stmt, book_title, book_authors, self:md5(book_stats.file)))
```
Comment (houqp) [inline plugins/statistics.koplugin/main.lua]: For dynamic queries, always use prepare then bind to execute it :) Doing manual string formatting is error prone because there are all kinds of weird keywords that need to be escaped to construct a valid query. Doing our own escaping would eventually end up with a reimplementation of bind, so might as well use bind from the start. bind should automatically escape the argument for you based on Lua types, so you won't need to do the tile and author gsub.

File: plugins/statistics.koplugin/main.lua (hunk 11 above)
Comment (houqp) [inline plugins/statistics.koplugin/main.lua]: In that case, wouldn't it be better to write it as `Pages: %1, Time: %2`?

File: plugins/statistics.koplugin/main.lua [hunk 13]
```diff
@@ -369,236 +789,282 @@ end
...
-                    end,
-                })
-                total_books_time = total_books_time + tonumber(book_result.total_time)
-            end
+function ReaderStatistics:getTotalStats()
+    self:insertDB(self.id_curr_book)
+    local conn = SQ3.open(db_location)
+    local sql_stmt = [[
+        SELECT sum(period)
+        FROM   page_sta
```
Comment (houqp) [inline plugins/statistics.koplugin/main.lua]: typo? shouldn't it be `page_stat`?

File: plugins/statistics.koplugin/main.lua [hunk 14]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
+
+function ReaderStatistics:insertDB(id_book)
+    self.pages_stats[TimeVal:now().sec] = self.curr_page
+    if id_book == nil or util.tableSize(self.pages_stats) < 2 then
+        return
+    end
+    local diff_time
+    local conn = SQ3.open(db_location)
+    local sorted_performance = {}
+    for k, v in pairs(self.pages_stats) do
```
Comment (houqp) [inline plugins/statistics.koplugin/main.lua]: when you are iterating through a table, it's better to give `k`, `v` specific names like `time` and `page`.

File: plugins/statistics.koplugin/main.lua [hunk 15]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
+        table.insert(sorted_performance, k)
+    end
+    table.sort(sorted_performance)
+    conn:exec('BEGIN')
+    local stmt = conn:prepare("INSERT OR IGNORE INTO page_stat VALUES(?, ?, ?, ?)")
+    for i=1, #sorted_performance - 1 do
+        diff_time = sorted_performance[i+1] - sorted_performance[i]
+        if diff_time <= self.page_max_read_sec and diff_time >= self.page_min_read_sec  then
```
Comment (houqp) [inline plugins/statistics.koplugin/main.lua]: this logic can be simplified to the following:

```lua
if diff_time >= self.page_min_read_sec then
    stmt:reset():bind(id_book,
                      self.pages_stats[sorted_performance[i]], 
                      sorted_performance[i], 
                      math.min(diff_time, self.page_max_read_sec)):step()
end

**Comment by houqp** [comment]
Sorry for the delay :( @apletnev and @Frenzie caught most of the issues. I added couple more comments. The rest looks mostly good to me, awesome work :+1:

File: plugins/statistics.koplugin/main.lua [hunk 16]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
+                end,
+            })
+        end
+    else  -- first time convertion to sqlite database
+        self.convert_to_db = true
+        if not conn:exec("pragma table_info('book');") then
+            if #ReadHistory.hist > 0 then
+                local info = InfoMessage:new{
+                    text =_("New version of statistics plugin is detected.\n" ..
+                        "You need to convert statistics data to database.\n"..
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: "You need" implies user action. This should either be changed to a confirm dialog (so that users can opt out of the process) or it should just say something like "Statistics data needs to be converted into the new database format."

File: plugins/statistics.koplugin/main.lua [hunk 17]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
-        table.insert(result, {
-            k,
-            T(_("Pages (%1) Time: %2"), v.count, util.secondsToClock(v.read, false))
-        })
+function ReaderStatistics:checkInitDatabase()
+    local conn = SQ3.open(db_location)
+    if self.convert_to_db then      -- if conversion to sqlite was doing earlier
+        if not conn:exec("pragma table_info('book');") then
+            UIManager:show(ConfirmBox:new{
+                text = T(_("Cannot open database in %1.\nProbably the database has been deleted or moved.\n " ..
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: The database may have been moved or deleted.

File: plugins/statistics.koplugin/main.lua [hunk 18]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
+                    "Do you want to create empty database?"), db_location),
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Do you want to create *an* empty database?

File: plugins/statistics.koplugin/main.lua [hunk 19]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
+            })
+        end
+    else  -- first time convertion to sqlite database
+        self.convert_to_db = true
+        if not conn:exec("pragma table_info('book');") then
+            if #ReadHistory.hist > 0 then
+                local info = InfoMessage:new{
+                    text =_("New version of statistics plugin is detected.\n" ..
+                        "You need to convert statistics data to database.\n"..
+                        "It make take a few minutes.\nPlease wait...")}
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Typo, "This *may* take a few minutes."

Also, please use a real ellipsis (`…`) instead of three periods.

File: plugins/statistics.koplugin/main.lua [hunk 20]
```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
...
+                UIManager:forceRePaint()
+                local nr_book = self:migrateToDB(conn)
+                UIManager:close(info)
+                UIManager:forceRePaint()
+                UIManager:show(InfoMessage:new{
+                    text =T(_("Convertion completed.\nImported %1 books to database."),nr_book), timeout = 4 })
+            else
+                self:createDB(conn)
+                UIManager:show(InfoMessage:new{
+                    text =_("New version of statistics plugin is detected.\n" ..
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Will this always show on the first run? It might be better to leave out the info message here.

**Comment by robert00s** [comment]
@houqp, @Frenzie 
Thanks for review. I'll try to make changes soon.

**Comment by robert00s** [comment]
Last commit adds "reset reading stats" options (reference #2920).

screenshot 2017-07-30 16-51-19

**Comment by AlanSP1** [comment]
Maybe it would be better if we have consistent language?

If question is if we want statistics reset, button should also have reset, not delete. Or vice versa, we can ask if we want statistics deleted, then button text would be consistent.

**Comment by robert00s** [comment]
You're right. I will correct it

**Comment by Frenzie** [comment]
Are there any remaining issues other than the conflict with frontend/apps/reader/modules/readerfooter.lua?

**Comment by robert00s** [comment]
@Frenzie 
There are no remaining issues other than the conflict.

**Comment by Frenzie** [comment]
Alright, then I'll merge it once that's resolved. :-)

**Comment by robert00s** [comment]
So we need to wait for @houqp

**Comment by Frenzie** [comment]
Oh, I thought you resolved @houqp 's comments? But no problem.

**Comment by robert00s** [comment]
Yes, I resolved @houqp's comments.

File: plugins/statistics.koplugin/main.lua [hunk 21]
```diff
@@ -919,32 +906,25 @@ function ReaderStatistics:getReadingProgressStats(sdays)
...
     local sql_stmt = [[
-        SELECT strftime('%%Y-%%m-%%d', start_time, 'unixepoch', 'localtime') AS dates,
-               count(DISTINCT page)                                          AS pages,
-               sum(period)                                                   AS periods,
-               start_time
-        FROM   page_stat
-        WHERE  start_time >= '%s'
-        GROUP  BY dates
-        ORDER  BY dates DESC
+            SELECT dates,
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Total nit but the indentation seems to be one too deep?

**Comment by robert00s** [comment]
I think it's ready to release.
I tested the plugin for several weeks and it worked fine.

**Comment by Frenzie** [comment]
:tada:

**Comment by robert00s** [comment]
Thanks :)

**Comment by poire-z** [comment]
Sorry, a bit late to mention this:
https://github.com/koreader/koreader/blob/c6cfed92d7123eef1c5aa28702d3eea3dcd34226/plugins/statistics.koplugin/main.lua#L23
Wouldn't it better to have it named `statistics.sqlite3` ? So it's obvious what it is. Or does it act as some more general book db ?
It will be in the settings dirrectory alongside the `bookinfo_cache.sqlite3` from coverbrowser plugin, so it may helps to avoid confusion.

**Comment by poire-z** [comment]
Got a crash: it happened I had statistics disabled (I don't remember having disable them, may be a tap bounce), and started a new book with them disabled. When going to see current book stats:
```
./luajit: frontend/ui/widget/keyvaluepage.lua:339: attempt to get length of field 'kv_pairs' (a nil value)
stack traceback:
 frontend/ui/widget/keyvaluepage.lua:339: in function 'init'
 frontend/ui/widget/widget.lua:48: in function 'new'
 plugins/statistics.koplugin/main.lua:622: in function 'callback'
 frontend/ui/widget/touchmenu.lua:632: in function 'action'
```

Quick solution to avoid the crash could be to return an empty dict instead of nothing in:
https://github.com/koreader/koreader/blob/a720eaceb50f32465e22d8210638cc6d2541f7d7/plugins/statistics.koplugin/main.lua#L767-L770
But may be you want to do something else if statistics are disabled/not available for this book, or the book is not found in db.

**Comment by robert00s** [comment]
@poire-z 
Thanks, I try to fix it.

**Comment by robert00s** [comment]
@Frenzie, @poire-z 
Do you know the method (trick) that "Current book" is enable when `self.is_enabled` is `true` and disabled when is `false`. But it should change while book is still open. That code doesn't work:
```
           {
                text = _("Current book"),
                callback = function()
                    UIManager:show(KeyValuePage:new{
                        title = _("Statistics"),
                        kv_pairs = self:getCurrentStat(self.id_curr_book),
                    })
                end,
                enabled = not self:isDocless() and self.is_enabled
            },
```
I mean that line
`enabled = not self:isDocless() and self.is_enabled`

`enabled` is calculated while plugin is starting.

**Comment by poire-z** [comment]
Must be something like:
```lua
enabled_func = function() return not self:isDocless() and self.is_enabled end,
```

**Comment by Frenzie** [comment]
What he said. ;-)

**Comment by robert00s** [comment]
Thanks a lot. That's what I needed. Works perfectly :)
